src/lines/MultiLine.sol

```diff
@@ -0,0 +1,190 @@
+    }
+
+    function _send(address, uint256 toChainId, address toDapp, bytes calldata message, bytes calldata params)
+        internal
+        override
+    {
+        RemoteCallArgs memory args = abi.decode(params, (RemoteCallArgs));
+        multiSend(
+            MultiSendArgs(
+                args.names,
+                toChainId,
+                toDapp,
+                message,
+                args.params,
+                args.fees,
+                args.salt,
+                args.expiration,
+                args.threshold
+            )
+        );
+    }
+
+    function multiSend(MultiSendArgs memory args) public payable {
+        require(args.toChainId != LOCAL_CHAINID(), "!toChainId");
+
+        address fromDapp = msg.sender;
+        LineMsg memory lineMsg = LineMsg({
+            fromChainId: LOCAL_CHAINID(),
+            toChainId: args.toChainId,
+            fromDapp: fromDapp,
+            toDapp: args.toDapp,
+            salt: args.salt,
+            message: args.message,
+            expiration: args.expiration,
+            threshold: args.threshold
+        });
+        bytes32 lineMsgId = hash(lineMsg);
+        bytes memory encoded = abi.encodeWithSelector(MultiLine.multiRecv.selector, lineMsg);
+
+        uint256 totalFee = 0;
+        for (uint256 i = 0; i < args.names.length; i++) {
+            string memory name = args.names[i];
+            uint256 fee = args.fees[i];
+            address line = REGISTRY.getLine(name);
+            require(line != address(0), "!name");
+            IMessageLine(line).send{value: fee}(args.toChainId, _toLine(args.toChainId), encoded, args.params[i]);
+            totalFee += fee;
+        }
+
+        require(totalFee == msg.value, "!fees");
+        emit LineMessageSent(lineMsgId, args.names, lineMsg);
+    }
+
+    function multiRecv(LineMsg calldata lineMsg) external payable {
+        address line = _msgLine();
+        require(REGISTRY.isTrustedLine(line), "!line");
+        uint256 fromChainId = _fromChainId();
+        require(LOCAL_CHAINID() == lineMsg.toChainId, "!toChainId");
+        require(fromChainId == lineMsg.fromChainId, "!fromChainId");
+        require(fromChainId != LOCAL_CHAINID(), "!fromChainId");
+        require(_xmsgSender() == _fromLine(fromChainId), "!xmsgSender");
+        bytes32 lineMsgId = hash(lineMsg);
+        require(deliverifyOf[lineMsgId][line] == false, "deliveried");
+        deliverifyOf[lineMsgId][line] = true;
+        ++countOf[lineMsgId];
+
+        emit LineMessageConfirmation(lineMsgId, ILineMetadata(line).name());
+
+        if (block.timestamp > lineMsg.expiration) {
+            emit LineMessageExpired(lineMsgId);
+            return;
+        }
+
+        require(doneOf[lineMsgId] == false, "done");
+        if (countOf[lineMsgId] >= lineMsg.threshold) {
+            doneOf[lineMsgId] = true;
+            _recv(lineMsg.fromChainId, lineMsg.fromDapp, lineMsg.toDapp, lineMsg.message);
```

Comment:
Before final execute and _recv, need to validate all the lineMsg or their hashes are the same.